### PR TITLE
Introduce OAS linting via Spectral

### PR DESCRIPTION
**Pull request by eiof**
[Spectral](https://meta.stoplight.io/docs/spectral/ZG9jOjYyMDc0Mg-concepts) is a linter that not only helps with linting our Open API spec, but also allows us to introduce rules for API documentation consistency and validity. This introduction of Spectral does not fix all of the current issues with our spec. It will help to prevent the addition of issues going forward.

* Spectral can be installed and used locally, but it's not a requirement for productivity, so there's no introduction of dependency management (package.json, install scripts, etc.)
* Spectral will be ran on PRs
* Spectral is producing warnings that we can address later

There are alternatives, including a [linter from Redocly](https://github.com/Redocly/redocly-cli), which is what we use for our documentation. I've tried Redocly's linter locally, and it generates roughly the same results. I don't have a strong stance on what tool is better, though I went with Spectral as it seems more mature at the moment.

### CODE REVIEW

File: openapi-v3.yaml [hunk 1]
```diff
@@ -86,7 +86,19 @@ paths:
...
+              parameters:
+                accountId: $response.body#/data/0/id
+            ListProjects:
+              operationId: listProjects
+              parameters:
+                accountId: $response.body#/data/0/id
+            ListJobsForAccount:
+              operationId: listJobsForAccount
+              parameters:
+                accountId: $response.body#/data/0/id
```
Comment (eiof) [inline openapi-v3.yaml:101]: Example of using links to enable API traversal

File: openapi-v3.yaml [hunk 2]
```diff
@@ -223,6 +247,9 @@ paths:
           name: order_by
           schema:
             type: string
+            anyOf:
+            - enum: [id, created_at, -id, -created_at]
+            - type: "string"
```
Comment (eiof) [inline openapi-v3.yaml:252]: We may want to consider removing or adding to this list. I believe any field on Job can be used for `order_by`, but these are the two fields I most commonly see.

File: openapi-v3.yaml [hunk 3]
```diff
@@ -724,16 +752,14 @@ paths:
...
-                  child_map: {}
-
+                example: "# package.yaml
+                  packages:
+                  - package: dbt-labs/audit_helper
+                  version: 0.5.0
+                  - package: dbt-labs/codegen
+                  version: 0.5.0
+                  - package: dbt-labs/dbt_utils
+                  version: 0.8.2"
```
Comment (eiof) [inline openapi-v3.yaml:762]: Linting fix, I chose to use an example package.yml as it's potentially more tangible, but we could go back to using an empty manifest if desired.

File: openapi-v3.yaml [hunk 4]
```diff
@@ -841,7 +868,8 @@ components:
           type: integer
           example: 5000
         dbt_project_subdirectory:
-          type: [null, integer]
+          type: string
+          nullable: true
```
Comment (eiof) [inline openapi-v3.yaml:872]: Linting & type fix, `nullable: true` is the correct way to specify that a property can be null

**Review by eiof — COMMENTED** [review]
Annotations for reviewers

File: openapi-v3.yaml (hunk 2 above)
Comment (gcs272) [inline openapi-v3.yaml:252]: I think this is redundant with the enum - at least redoc displays it as two separate options:

Screen Shot 2022-06-03 at 11 54 01 AM

**Review by gcs272 — COMMENTED** [review]
Looks great, and I love that the warnings get annotated in the diff. Maybe way down the line we'll want to fail CI if the warnings increase or something like that?

File: openapi-v3.yaml (hunk 1 above)
Comment (gcs272) [reply openapi-v3.yaml:101]: FYI - super helpful, but I don't see anything come through when rendering via redoc

File: openapi-v3.yaml (hunk 2 above)
Comment (gcs272) [reply openapi-v3.yaml:252]: I think this is the most correct thing though (or just `string` with examples) - I missed your comment earlier about there being other options and these just being the most common

Comment (eiof) [reply openapi-v3.yaml:252]: I'm going to remove this for now since it's an incomplete list anyways.

File: openapi-v3.yaml (hunk 1 above)
Comment (eiof) [reply openapi-v3.yaml:101]: Yeah, unfortunately Redocly nor Postman do not seem to do anything with this information at this time. However, other tools like [schemathesis can use this to generate a sequence of operations to test](https://schemathesis.readthedocs.io/en/stable/stateful.html).

Comment (kaitlinr) [reply openapi-v3.yaml:101]: Ah, interesting!  Is the recommendation here that we replace our Redocly usage with Spectral?  Do we know all the places in which Redocly is currently used?

**Comment by kaitlinr** [comment]
> Looks great, and I love that the warnings get annotated in the diff. Maybe way down the line we'll want to fail CI if the warnings increase or something like that?

Personally, I am all about this idea.  It's an easy way to enforce an API contract without a bunch of manual effort.

**Comment by eiof** [comment]
@kaitlinr I don't understand ownership and buy-in well either. @jeremyhutt11 and @gcs272 might lead several changes to this repository in an effort to improve our API documentation, so we should probably figure this out!

I may have caused some confusion with Redocly. There are two tools
1. [Redocly Reference](https://redocly.com/reference/): as far as I know, we use this, and will continue to use this for our API's documentation layout
2. [redocly-cli](https://redocly.com/redocly-cli/): we don't use this; this tool provides linter and deployment workflows, so it compliments Redocly Reference

So there are no deprecations here, only additions!

File: openapi-v3.yaml (hunk 1 above)
Comment (eiof) [reply openapi-v3.yaml:101]: Spectral will complement Redocly Reference, which is used for our API documentation. We won't be replacing Redocly

**Review by gcs272 — APPROVED** [review]
This looks awesome! Given the low(-ish) edit volume, I think my vote is to merge as is

**Review by Topherhindman — APPROVED** [review]

**Comment by kaitlinr** [comment]
> @kaitlinr I don't understand ownership and buy-in well either. @jeremyhutt11 and @gcs272 might lead several changes to this repository in an effort to improve our API documentation, so we should probably figure this out!
> 
> I may have caused some confusion with Redocly. There are two tools
> 
> 1. [Redocly Reference](https://redocly.com/reference/): as far as I know, we use this, and will continue to use this for our API's documentation layout
> 2. [redocly-cli](https://redocly.com/redocly-cli/): we don't use this; this tool provides linter and deployment workflows, so it compliments Redocly Reference -- this is the alternative tool that I was considering
> 
> So there are no deprecations here, only additions!

Ah, thank you for breaking this down for me Matt!  I was thinking that we already had `redocly-cli` because we were using `Redocly Reference`.  That was my mistake.  I'm tracking now! And after reading up more on Spectral, I'm all good here.

**Comment by kaitlinr** [comment]
> I don't understand ownership and buy-in well either.

I think Graham is on to something here given the low edit volume.  I like that you've been updating the glossary and adding new tools / frameworks as you're going along.  I think it might be worth including these new tools / ideas in the update we post to #tmp-dev-api-services-strategy.  If there's pushback, then we can consider a more public post in #dev or formulating an RFC.  Does that seem like a reasonable approach?

**Pull request merged**